mypy/modulefinder.py

```diff
@@ -124,6 +154,39 @@ def clear(self) -> None:
         self.initial_components.clear()
         self.ns_ancestors.clear()
 
+    def find_module_via_source_set(self, id: str) -> Optional[ModuleSearchResult]:
+        if not self.source_set:
+            return None
+        p = self.source_set.source_modules.get(id, None)
+        if p and self.fscache.isfile(p):
+            # NB: need to make sure we still have __init__.py all the way up
+            # otherwise we might have false positives compared to slow path
+            d = os.path.dirname(p)
+            for i in range(id.count('.')):
+                if not self.fscache.isfile(os.path.join(d, '__init__.py')):
+                    return None
+                d = os.path.dirname(d)
+            return p
+
+        idx = id.rfind('.')
+        if idx != - 1:
+            parent = self.find_module_via_source_set(id[:idx])
+            if (
+                    parent and isinstance(parent, str)
+                    and not parent.endswith('__init__.py')
+                    and not self.fscache.isdir(os.path.splitext(parent)[0])
+            ):
+                # if
+                #   1. we're looking for foo.bar.baz
+                #   2. foo.bar.py[i] is in the source set
+                #   3. foo.bar is not a directory
+                # then we don't want to go spelunking in other search paths to find
+                # another 'bar' module, because it's a waste of time and even in the
+                # unlikely event that we did find one that matched, it probably would
+                # be completely unrelated and undesirable
+                return ModuleNotFoundReason.NOT_FOUND
```

Comment:
this is probably the most controversial behavior change, but unfortunately,also the key ingredient in achieving any measurable speedup, because most calls to `find_module` will fail, e.g. `from typing import Dict` triggers a call to `find_module('typing.Dict')`...